src/serprog.c

```diff
@@ -0,0 +1,519 @@
+}
+
+static int serprog_initialize(const PROGRAMMER *pgm, const AVRPART *part) {
+  if(part->prog_modes & PM_TPI) {
+    // We do not support TPI; this is a dedicated SPI thing
+    pmsg_error("the %s programmer does not support TPI\n", pgmid);
+    return -1;
+  }
+
+  unsigned char buf[32];
+
+  // Set SPI clock frequency
+  if(is_serprog_cmd_supported(my.cmd_bitmap, S_CMD_S_SPI_FREQ)) {
+    memset(buf, 0, sizeof buf);
+    uint32_t frequency =
+      pgm->bitclock > 0? 1/pgm->bitclock:
+      part->factory_fcpu > 0? part->factory_fcpu/4:
+      250000;
+    write_le32(buf, frequency);
+    if(perform_serprog_cmd(pgm, S_CMD_S_SPI_FREQ, buf, 4, buf, 4) != 0) {
+      pmsg_error("cannot set SPI frequency %u Hz\n", (unsigned) frequency);
+      return -1;
+    }
+    my.actual_frequency = read_le32(buf);
+  }
+
+  // Set active chip select
+  if(is_serprog_cmd_supported(my.cmd_bitmap, S_CMD_S_SPI_CS)) {
+    memset(buf, 0, sizeof buf);
+    buf[0] = my.cs;
+    if(perform_serprog_cmd(pgm, S_CMD_S_SPI_CS, buf, 1, NULL, 0) != 0) {
+      pmsg_error("cannot change CS\n");
+      return -1;
+    }
+  }
+
+  // Set full duplex
+  memset(buf, 0, sizeof buf);
+  buf[0] = SPI_MODE_FULL_DUPLEX;
+  if(perform_serprog_cmd(pgm, S_CMD_S_SPI_MODE, buf, 1, NULL, 0) != 0) {
+    pmsg_error("cannot set SPI full duplex mode\n");
+    return -1;
+  }
+
+  // Set output
+  if(is_serprog_cmd_supported(my.cmd_bitmap, S_CMD_S_PIN_STATE)) {
+    memset(buf, 0, sizeof buf);
+    buf[0] = 1;                 // Pin state enable
+    if(perform_serprog_cmd(pgm, S_CMD_S_PIN_STATE, buf, 1, NULL, 0) != 0) {
+      pmsg_error("cannot enable pin state\n");
+      return -1;
+    }
+  }
+
+  // Enable the CS/reset pin
+  const unsigned char cs_mode = CS_MODE_SELECTED;
+
+  if(perform_serprog_cmd(pgm, S_CMD_S_CS_MODE, &cs_mode, 1, NULL, 0) != 0) {
+    pmsg_error("cannot enable the reset pin\n");
+    return -1;
+  }
+
+  int tries, ret;
+
+  // Enable programming on the part
+  tries = 0;
+  do {
+    ret = pgm->program_enable(pgm, part);
+    if(ret == 0 || ret == -1)
+      break;
+  } while(tries++ < 65);
+
+  if(ret)
+    pmsg_error("AVR device not responding\n");
+
+  return ret;
+}
+
+/* used linuxspi.c as a template:
+ *   Copyright (C) 2013 Kevin Cuzner <kevin@kevincuzner.com>
+ *   Copyright (C) 2018 Ralf Ramsauer <ralf@vmexit.de>
+ */
+static int serprog_program_enable(const PROGRAMMER *pgm, const AVRPART *p) {
+  unsigned char cmd[4], res[4];
+
+  if(!p->op[AVR_OP_PGM_ENABLE]) {
+    pmsg_error("program enable instruction not defined for part %s\n", p->desc);
+    return -1;
+  }
+
+  memset(cmd, 0, sizeof cmd);
+  avr_set_bits(p->op[AVR_OP_PGM_ENABLE], cmd);  // Set the cmd
+  pgm->cmd(pgm, cmd, res);
+
+  if(res[2] != cmd[1]) {
+
+    /** From ATtiny441 datasheet:
+     *
+     * In some systems, the programmer cannot guarantee that SCK is held low
+     * during power-up. In this case, RESET must be given a positive pulse after
```

Comment:
We know the programmer here. It's a hardware that conforms to the Serprog protocol. Does that guarantee SCK being held low? In which case this `if()` branch would *never* be executed, I suppose, and the loop in `initialize()` could simply be replaced by
```
  if(pgm->program_enable(pgm, part))
    pgm_error("...");
```
That would be the standard way to end `initialize()`. If that can be safely built back you don't need the ref to the `linuxspi` authors either. Your call @funkeleinhorn